### PR TITLE
Add Unit Test for TweetFetcher

### PR DESCRIPTION
**Pull request by Max-Leopold**

### CODE REVIEW

File: twitterpuller/src/main/java/de/brandwatch/minianalytics/twitterpuller/TwitterPullerApplication.java [hunk 1]
```diff
@@ -15,12 +15,18 @@ public static void main(String[] args) {
         SpringApplication.run(TwitterPullerApplication.class, args);
     }
 
+    private TweetFetcher tweetFetcher;
+
+    @Autowired
+    Producer producer;
```
Comment (YNedderhoff) [inline twitterpuller/src/main/java/de/brandwatch/minianalytics/twitterpuller/TwitterPullerApplication.java]: This can be private. Also I would recommend to get into the habit of using constructor based Autowiring instead, which is generally [recommended](https://kinbiko.com/java/dependency-injection-patterns/). Then it can also be `final`.

File: twitterpuller/src/main/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcher.java [hunk 2]
```diff
@@ -1,51 +1,25 @@
...
+import org.springframework.stereotype.Component;
 import twitter4j.*;
 
-import java.time.Instant;
 
+@Component
 public class TweetFetcher {
 
-    public TwitterStream getTwitterStream(Producer producer) {
+    private TwitterPullerStatusListener listener;
```
Comment (YNedderhoff) [inline twitterpuller/src/main/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcher.java]: can be final

File: twitterpuller/src/main/java/de/brandwatch/minianalytics/twitterpuller/twitter/TwitterPullerStatusListener.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import twitter4j.Status;
+import twitter4j.StatusDeletionNotice;
+import twitter4j.StatusListener;
+
+import java.time.Instant;
+
+@Component
+public class TwitterPullerStatusListener implements StatusListener {
+
+    private Producer producer;
```
Comment (YNedderhoff) [inline twitterpuller/src/main/java/de/brandwatch/minianalytics/twitterpuller/twitter/TwitterPullerStatusListener.java]: can be final

File: twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import twitter4j.User;
+
+import java.util.Date;
+
+
+@RunWith(MockitoJUnitRunner.class)
+public class TweetFetcherTest {
+
+    @Mock
+    Producer producer;
```
Comment (YNedderhoff) [inline twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java]: can be private

File: twitterpuller/pom.xml [hunk 5]
```diff
@@ -22,6 +22,22 @@
             <groupId>org.twitter4j</groupId>
             <artifactId>twitter4j-stream</artifactId>
         </dependency>
+        <dependency>
```
Comment (YNedderhoff) [inline twitterpuller/pom.xml]: You'll want to add these to the dependency management section of your parent pom

File: twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+import org.mockito.junit.MockitoJUnitRunner;
+import twitter4j.Status;
+import twitter4j.User;
+
+import java.util.Date;
+
+
+@RunWith(MockitoJUnitRunner.class)
+public class TweetFetcherTest {
```
Comment (YNedderhoff) [inline twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java]: It's the test for the StatusListener, not the TweetFetcher

File: twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        Date date = new Date();
+
+        when(status.getText()).thenReturn("Hello World");
+        when(status.getUser().getName()).thenReturn("Max Leopold");
+        when(status.getCreatedAt()).thenReturn(date);
+
+        ArgumentCaptor<Resource> argumentCaptor = ArgumentCaptor.forClass(Resource.class);
+
+        TwitterPullerStatusListener twitterPullerStatusListener = new TwitterPullerStatusListener(producer);
+        TweetFetcher tweetFetcher = new TweetFetcher(twitterPullerStatusListener);
```
Comment (YNedderhoff) [inline twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java]: This seems unnecessary?

File: twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+        Date date = new Date();
+
+        when(status.getText()).thenReturn("Hello World");
+        when(status.getUser().getName()).thenReturn("Max Leopold");
+        when(status.getCreatedAt()).thenReturn(date);
+
+        ArgumentCaptor<Resource> argumentCaptor = ArgumentCaptor.forClass(Resource.class);
+
+        TwitterPullerStatusListener twitterPullerStatusListener = new TwitterPullerStatusListener(producer);
```
Comment (YNedderhoff) [inline twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java]: Instead of this, you should be able to do 

```java
@InjectMocks
private TwitterPullerStatusListener listener;
```

on top, unless you have a good reason to have a separate instance per test.

File: twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        TweetFetcher tweetFetcher = new TweetFetcher(twitterPullerStatusListener);
+
+        twitterPullerStatusListener.onStatus(status);
+
+        verify(producer).send(argumentCaptor.capture());
+
+        assertEquals("Hello World", argumentCaptor.getValue().getText());
```
Comment (YNedderhoff) [inline twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java]: This is one way to do it. You should alternatively also be able to create a `expectedResource` object in this test, and then just do one `verify(producer).send(eq(expectedResource));`

File: twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+@RunWith(MockitoJUnitRunner.class)
+public class TweetFetcherTest {
+
+    @Mock
+    Producer producer;
+
+    @Test
+    public void testIfResourceObjectIsGeneratedCorrectly() {
+
+        Status status = mock(Status.class, RETURNS_DEEP_STUBS);
```
Comment (YNedderhoff) [inline twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java]: `RETURNS_DEEP_STUBS` is not considered to be good, clean style:

```
This feature should rarely be required for regular clean code! Leave it for legacy code.
     * Mocking a mock to return a mock, to return a mock, (...), to return something meaningful
     * hints at violation of Law of Demeter or mocking a value object (a well known anti-pattern).
```

Instead, below, you should mock a `User` independently.

I have the feeling that this doesn't even exist in mockito 3, although I am not 100% sure. But you seem to use an older version (2.x.x), which you should fix as well.

File: twitterpuller/src/test/resources/application.properties [hunk 11]
```diff
@@ -0,0 +1 @@
+kafka.bootstrap-servers
```
Comment (YNedderhoff) [inline twitterpuller/src/test/resources/application.properties]: Why is this added in this PR?

Comment (Max-Leopold) [reply twitterpuller/src/test/resources/application.properties]: I'ts from another Test Case I was working on. Its wrong on this commit.

File: twitterpuller/pom.xml (hunk 5 above)
Comment (Max-Leopold) [reply twitterpuller/pom.xml]: They are already managed through the dependency Management of the Spring parent pom.

File: twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java (hunk 9 above)
Comment (Max-Leopold) [reply twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java]: This doesn't work. 
```
producer.send(
    de.brandwatch.minianalytics.twitterpuller.model.Resource@452e19ca
);
-> at de.brandwatch.minianalytics.twitterpuller.twitter.StatusListenerTest.testIfResourceObjectIsGeneratedCorrectly(StatusListenerTest.java:62)

However, there was exactly 1 interaction with this mock:
producer.send(
    de.brandwatch.minianalytics.twitterpuller.model.Resource@1ec9bd38
);
-> at de.brandwatch.minianalytics.twitterpuller.twitter.TwitterPullerStatusListener.onStatus(TwitterPullerStatusListener.java:35)
```
But i could do something pretty similar:

`assertTrue(new ReflectionEquals(expectedResource).matches(argumentCaptor.getValue()));`
This does work.

Comment (Max-Leopold) [reply twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java]: What do you prefer?

File: twitterpuller/pom.xml (hunk 5 above)
Comment (YNedderhoff) [reply twitterpuller/pom.xml]: All of them? Interesting - but I recommend overriding that with your own versions, specifically Mockito 3, junit 5.

File: twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java (hunk 9 above)
Comment (YNedderhoff) [reply twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java]: It will probably work if you implement `equals` and `hashCode` on your `Resource`? What IntelliJ generates for you is likely enough. It is currently comparing only the object hashes, which are obviously different. You want it to compare by the field values instead.

https://stackoverflow.com/questions/2265503/why-do-i-need-to-override-the-equals-and-hashcode-methods-in-java

Comment (Max-Leopold) [reply twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java]: If I implement what IntelliJ generates, it just calls `super.equals(obj) `. And with that it still fails...

File: twitterpuller/pom.xml (hunk 5 above)
Comment (Max-Leopold) [reply twitterpuller/pom.xml]: Its currently Mockito version 2.23.4 and JUnit version 4.12 so I will upgrade that.

File: twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java (hunk 9 above)
Comment (YNedderhoff) [reply twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java]: Really? That's odd, if I do it it generates this:

```
@Override
public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof Resource)) return false;
    Resource resource = (Resource) o;
    return Objects.equal(getAuthor(), resource.getAuthor()) &&
            Objects.equal(getText(), resource.getText()) &&
            Objects.equal(getDate(), resource.getDate());
}

@Override
public int hashCode() {
    return Objects.hashCode(getAuthor(), getText(), getDate());
}
```

You'll depend on [Guava](https://mvnrepository.com/artifact/com.google.guava/guava/28.1-jre) for that though:

Comment (Max-Leopold) [reply twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java]: Oh, I see why. You can just type `equals` into your Code and IntelliJ generates an equals method like I have. But if you go cmd+m and then generate hashCode and equals you get the methods like you have... didn't knew they were different...

Comment (Max-Leopold) [reply twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcherTest.java]: With these hashCode/equal Methods `verify(producer).send(eq(expectedResource));` does work.

File: pom.xml [hunk 12]
```diff
@@ -39,6 +39,16 @@
                 <artifactId>twitter4j-stream</artifactId>
                 <version>4.0.6</version>
             </dependency>
+            <dependency>
+                <groupId>org.mockito</groupId>
+                <artifactId>mockito-junit-jupiter</artifactId>
```
Comment (YNedderhoff) [inline pom.xml]: There's v3: https://mvnrepository.com/artifact/org.mockito/mockito-junit-jupiter/3.0.0

File: twitterpuller/pom.xml [hunk 13]
```diff
@@ -38,6 +58,17 @@
...
                 </configuration>
             </plugin>
+            <plugin>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <version>2.19.1</version>
+                <dependencies>
+                    <dependency>
+                        <groupId>org.junit.platform</groupId>
+                        <artifactId>junit-platform-surefire-provider</artifactId>
+                        <version>1.0.1</version>
```
Comment (YNedderhoff) [inline twitterpuller/pom.xml]: There's a newer version: https://mvnrepository.com/artifact/org.junit.platform/junit-platform-surefire-provider/1.3.2

File: twitterpuller/src/main/java/de/brandwatch/minianalytics/twitterpuller/TwitterPullerApplication.java [hunk 14]
```diff
@@ -15,12 +14,15 @@ public static void main(String[] args) {
         SpringApplication.run(TwitterPullerApplication.class, args);
     }
 
+    private TweetFetcher tweetFetcher;
```
Comment (YNedderhoff) [inline twitterpuller/src/main/java/de/brandwatch/minianalytics/twitterpuller/TwitterPullerApplication.java]: canbe final

File: twitterpuller/src/main/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcher.java [hunk 15]
```diff
@@ -1,51 +1,24 @@
...
-            public void onStallWarning(StallWarning stallWarning) {
-
-            }
-
-            public void onException(Exception e) {
-                e.printStackTrace();
-            }
-        });
-
+    public TwitterStream getTwitterStream() {
```
Comment (YNedderhoff) [inline twitterpuller/src/main/java/de/brandwatch/minianalytics/twitterpuller/twitter/TweetFetcher.java]: Looking at it like this, now it would be even more straightforward to define as a simple bean, no?

```java
@Bean
public twitterStream(TwitterPullerStatusListener listener){
    TwitterStream twitterStream = TwitterStreamFactory.getSingleton();
    twitterStream.addListener(listener);
    return twitterStream;
}

Then you wouldn't need this class at all.

File: twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/StatusListenerTest.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    void testIfResourceObjectIsGeneratedCorrectly() {
+
+        Date date = new Date();
+
+        when(status.getText()).thenReturn("Hello World");
+        when(status.getUser()).thenReturn(user);
+        when(user.getName()).thenReturn("Max Leopold");
+        when(status.getCreatedAt()).thenReturn(date);
+
+        ArgumentCaptor<Resource> argumentCaptor = ArgumentCaptor.forClass(Resource.class);
```
Comment (YNedderhoff) [inline twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/StatusListenerTest.java]: I don't think this is used now?

**Comment by Max-Leopold** [comment]
@YNedderhoff All done.

File: twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/StatusListenerTest.java [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+import twitter4j.User;
+
+import java.util.Date;
+
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class StatusListenerTest {
```
Comment (YNedderhoff) [inline twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/StatusListenerTest.java]: Should be `TwitterPullerStatusListener`

File: twitterpuller/src/main/java/de/brandwatch/minianalytics/twitterpuller/TwitterPullerApplication.java [hunk 18]
```diff
@@ -15,12 +17,22 @@ public static void main(String[] args) {
         SpringApplication.run(TwitterPullerApplication.class, args);
     }
 
+    private final TwitterPullerStatusListener listener;
+
     @Autowired
-    private Producer producer;
+    public TwitterPullerApplication(TwitterPullerStatusListener listener) {
```
Comment (YNedderhoff) [inline twitterpuller/src/main/java/de/brandwatch/minianalytics/twitterpuller/TwitterPullerApplication.java]: No need to Autowire this. When you create a Bean with a param as you do below with the listener, it's automatically assumed to be a Bean as well and hence Autowired.

Also, you are calling the Bean method instead of using the Bean itself. I'd actually suggest to move away from using `PostConstruct` with a blocking method like your stream, as it will prevent your app to fully start. It will forever be stuck in the post construct phase.

I'd recommend just this:

```java
@Bean
public TwitterStream twitterStream(TwitterPullerStatusListener listener) {
    TwitterStream twitterStream = TwitterStreamFactory.getSingleton();
    twitterStream.addListener(listener);
    return twitterStream;
}


@Bean(initMethod = "sendTweetsToKafka")
public TweetProducer tweetProducer(TwitterStream twitterStream){
    return new TweetProducer(twitterStream);
}
```
together with a class `TweetProducer`:

```java
package de.brandwatch.minianalytics.twitterpuller.twitter;

import twitter4j.TwitterStream;

public class TweetProducer {

    private TwitterStream twitterStream;

    public TweetProducer(TwitterStream twitterStream) {
        this.twitterStream = twitterStream;
    }

    public void sendTweetsToKafka() {
        twitterStream.sample("de");
    }
}
```

I haven't tried it, but this should create the new `TweetProducer` as a bean and call the init method (`sendTweetsToKafka`) without  blocking anything.

**Comment by Max-Leopold** [comment]
All done.

File: twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TwitterPullerStatusListenerTest.java [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+class TwitterPullerStatusListenerTest {
+
+    @Mock
+    private Producer producer;
+
+    @Mock
+    private User user;
+
+    @Mock
+    Status status;
```
Comment (YNedderhoff) [inline twitterpuller/src/test/java/de/brandwatch/minianalytics/twitterpuller/twitter/TwitterPullerStatusListenerTest.java]: can be private

**Comment by Max-Leopold** [comment]
I'm currently trying to test the Kafka producer and consumer...easier said then done...
